### PR TITLE
BUG: Incorrect parsing of ISO 8601 durations strings in Timedelta constructor

### LINKED ISSUE #51882: BUG: Incorrect parsing of ISO 8601 durations strings in Timedelta constructor
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
import pandas as pd

# Raises an error for >2 digit hours or minutes
pd.Timedelta('PT100H')
pd.Timedelta('PT100M')

# Incorrectly parses any decimal component apart from seconds
pd.Timedelta('PT3.5H') == pd.Timedelta('PT5H3S')

# Raises an error for comma as decimal separator
pd.Timedelta('PT3,5H')

# Does not raise an error for plusses in numbers
pd.Timedelta('P+1S') == pd.Timedelta('PT1S')

# Does not raise an error for unsupported month component, instead parses it as minutes
pd.Timedelta('P1M') == pd.Timedelta('PT1M')

# Does not raise an error for hour and second components before the 'T' seperator
pd.Timedelta('P1H') == pd.Timedelta('PT1H')
pd.Timedelta('P1S') == pd.Timedelta('PT1S')

# Does not raise an error for day components after the 'T' seperator
pd.Timedelta('PT1D') == pd.Timedelta('P1D')

# Does not raise an error for duplicate components (and produces unexpected result)
pd.Timedelta('PT1S1S1S1S') == pd.Timedelta('PT20M34S')

# Does not raise an error for components in the incorrect order
pd.Timedelta('PT1M1H') == pd.Timedelta('PT1H1M')

# Does not raise an error for trailing 'T' separator
pd.Timedelta('P1DT') == pd.Timedelta('P1D')

# Does not raise an error for duplicate 'T' separators
pd.Timedelta('PTTT1H') == pd.Timedelta('PT1H')

# Does not raise an error for 'P' or 'T' in any position
pd.Timedelta('PTP1TPTH') == pd.Timedelta('PT1H')
```

### Numbers
- Raises a ValueError when the hours or minutes part is longer than 2 digits. There is nothing in the standard that requires this. There is an explicit check for the length of the hours and minute parts in the parser. #51820 https://github.com/pandas-dev/pandas/blob/f9be5d90a6bb72f2d11f47046f483efbcd5a787e/pandas/_libs/tslibs/timedeltas.pyx#L883
- Incorrectly parses all decimal components aparts from seconds.
- Does not allow the comma `,` as a decimal separator
- Allows plusses (`+`) to prefix numbers, this is outwith the standard. https://github.com/pandas-dev/pandas/blob/f9be5d90a6bb72f2d11f47046f483efbcd5a787e/pandas/_libs/tslibs/timedeltas.pyx#L880

### Components
- Parses month (`M`) components as minutes. https://github.com/pandas-dev/pandas/blob/f9be5d90a6bb72f2d11f47046f483efbcd5a787e/pandas/_libs/tslibs/timedeltas.pyx#L885
- Parses hour (`H`) or second (`S`) components before the `T`
- Parses day components after the `T`
- Parses duplicate components with weird results
- Parses components in the incorrect order

### `P` and `T` symbols
- Ignores `P` and `T` symbols anywhere in the string. https://github.com/pandas-dev/pandas/blob/f9be5d90a6bb72f2d11f47046f483efbcd5a787e/pandas/_libs/tslibs/timedeltas.pyx#L873

### Expected Behavior

The follwing should all parse without errors.
```python
pd.Timedelta("PT100H") == pd.Timedelta(hours=100)
pd.Timedelta("PT100M") == pd.Timedelta(minutes=100)
pd.Timedelta("PT3,5H") == pd.Timedelta(hours=3.5)
```
The follwing should all raise errors.
```python
# Bad numbers
pd.Timedelta("PT.1S")
pd.Timedelta("PT+1S")

# Bad components
pd.Timedelta("P1M") 
pd.Timedelta("P1H")
pd.Timedelta("P1S")
pd.Timedelta("PT1D")
pd.Timedelta("PT1M1H")
pd.Timedelta("P1S1S")

# Misplaced P or T
pd.Timedelta("P1PD")
pd.Timedelta("P1TD")
pd.Timedelta("P1DT")
```

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 2e218d10984e9919f0296931d92ea851c6a6faf5
python           : 3.11.1.final.0
python-bits      : 64

pandas           : 1.5.3
numpy            : 1.24.2
pytz             : 2022.7.1
dateutil         : 2.8.2
setuptools       : 65.5.0
pip              : 22.3.1
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
brotli           : None
fastparquet      : None
fsspec           : None
gcsfs            : None
matplotlib       : None
numba            : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : None
snappy           : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
zstandard        : None
tzdata           : None

</details>

### PR DESCRIPTION
**Pull request by rosstitmarsh**
- [x] Closes #51882
- [x] Closes #48122
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] ~~Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.~~
- [x] Added an entry in the latest `doc/source/whatsnew/v2.1.0.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/_libs/tslibs/timedeltas.pyx [hunk 1]
```diff
@@ -836,93 +837,45 @@ cdef int64_t parse_iso_format_string(str ts) except? -1:
...
-        ts = ts[1:]
-
-    for c in ts:
-        # number (ascii codes)
-        if 48 <= ord(c) <= 57:
+        int sign = 1
+        object sign_part, integer_part, fractional_part
+        list groups = [], units = []
+
+    iso_regex = re.compile(
```
Comment (WillAyd) [inline pandas/_libs/tslibs/timedeltas.pyx:845]: Can you annotate this? I like the idea of using a regex here but this is hard to digest

Comment (WillAyd) [reply pandas/_libs/tslibs/timedeltas.pyx:848]: Sorry should be more clear on ask. I think you should use the `re.VERBOSE` flag and pair that with comments

File: pandas/_libs/tslibs/timedeltas.pyx [hunk 2]
```diff
@@ -842,37 +842,42 @@ cdef int64_t parse_iso_format_string(str ts) except? -1:
...
-        r"^(-?)P(?!$)(?:"
-        r"(?:(-?)(\d+)(?:[.,](\d{1,9})(?=Y$))?Y)?"
-        r"(?:(-?)(\d+)(?:[.,](\d{1,9})(?=M$))?M)?"
-        r"(?:(-?)(\d+)(?:[.,](\d{1,9})(?=W$))?W)?"
-        r"(?:(-?)(\d+)(?:[.,](\d{1,9})(?=D$))?D)?"
+        # Allow starting with 'P' or '-P', disallow ending on 'P'.
+        r"^-?P(?!$)(?:"
+        # Capture the sign, integer part, and fractional part (but only if it
+        # is the final component).
+        r"(?:(-?)(\d+)(?:[.,](\d{1,9})(?=Y$))?Y)?"  # Years
```
Comment (WillAyd) [inline pandas/_libs/tslibs/timedeltas.pyx:853]: Should definitely tighten up these repetitions - `\d{1,9}` is not really necessary

File: pandas/_libs/tslibs/timedeltas.pyx (hunk 1 above)
Comment (rosstitmarsh) [reply pandas/_libs/tslibs/timedeltas.pyx]: It is now broken up with `re.VERBOSE` and commented.

**Comment by mroeschke** [comment]
Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in the main branch, address any review comments and/or failing tests, and we can reopen.

**Comment by rosstitmarsh** [comment]
@mroeschke I would like to reopen this pull request, I have updated it to the main branch. Issues #51882 and #48122 are still not resolved.

**Comment by MarcoGorelli** [comment]
thanks for your PR

I do feel a bit nervous about this (as I do about the original implementation, tbh...), and think this might be a good case for a parametric test

would it be possible to add some reference implementation, such as [isoduration](https://github.com/bolsote/isoduration), and then to use [hypothesis](https://github.com/HypothesisWorks/hypothesis) to generate random tests in which you assert that the output matches that of the reference implementation?

**Comment by rosstitmarsh** [comment]
> would it be possible to add some reference implementation, such as [isoduration](https://github.com/bolsote/isoduration), and then to use [hypothesis](https://github.com/HypothesisWorks/hypothesis) to generate random tests in which you assert that the output matches that of the reference implementation?

@MarcoGorelli I agree that is a good idea.
I am struggling with isoduration because [it creates duration strings where the number parts can be in scientific notation ](https://github.com/bolsote/isoduration/blob/ae0bd61507aa04e032402566a2c7c01a0402929c/src/isoduration/formatter/formatting.py#L34), this does not seem to be part of ISO8601

**Comment by MarcoGorelli** [comment]
thanks - maybe there's something else we can use then? will take a look

**Comment by mroeschke** [comment]
Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in the main branch, address any review comments and/or failing tests, and we can reopen.

**Pull request closed without merging**